File: Router/src/Modules/Hardware/Installation.cpp
#include "pch_App.h"

#include "Installation.h"
#include "../App.h"

#include "../Utils.h"

using namespace msgpack11;

namespace Modules {
	namespace Hardware {
		//----------
		Installation::Installation()
		{
			this->panel = ofxCvGui::Panels::makeWidgets();
			this->massFWUpdate = make_shared<MassFWUpdate>();
		}

		//----------
		Installation::~Installation()
		{

		}

		//----------
		string
			Installation::getTypeName() const
		{
			return "Hardware::Installation";
		}

		//----------
		void
			Installation::init()
		{
			this->onPopulateInspector += [this](ofxCvGui::InspectArguments& args) {
				this->populateInspector(args);
				};
			this->massFWUpdate->init();
		}

		//----------
		void
			Installation::update()
		{
			if (this->parameters.image.enabled.get()) {
				this->takeImage();
			}

			this->transmitFrame();

			if (this->needsRebuildColumns) {
				this->rebuildColumns();
			}

			// update columns
			for (const auto& column : this->columns) {
				column->update();
			}

			if (this->needsRebuildPanel) {
				this->rebuildPanel();
			}
		}

		//----------
		void
			Installation::deserialise(const nlohmann::json& json)
		{
			// Get the arrangement
			{
				if (json.contains("arrangement")) {
					Utils::deserialize(json["arrangement"], this->parameters.arrangement.columns);
					Utils::deserialize(json["arrangement"], this->parameters.arrangement.rows);
					Utils::deserialize(json["arrangement"], this->parameters.arrangement.columnWidth);
					Utils::deserialize(json["arrangement"], this->parameters.arrangement.flipped);
				}
			}

			this->rebuildColumns();

			// Deserialise settings for the columns themselves
			if (json.contains("columns")) {
				auto jsonColumns = json["columns"];

				const auto jsonColumnSettings = json.contains("columnCommonSettings")
					? json["columnCommonSettings"]
					: nlohmann::json();

				auto count = min(this->columns.size(), jsonColumns.size());

				for (size_t i = 0; i < count; i++) {
					auto jsonColumn = jsonColumns[i];
					auto column = this->columns[i];

					// Merge default settings with column settings
					jsonColumn.update(jsonColumnSettings);
					column->deserialise(jsonColumn);
				}
			}

			ofxCvGui::refreshInspector(this);
		}

		//----------
		void
			Installation::rebuildColumns()
		{
			if (!ofxCvGui::isBeingInspected(this)) {
				// In case we are selecting something inside a column
				ofxCvGui::inspect(nullptr);
			}

			this->columns.clear();

			Column::Settings columnSettings;
			{
				columnSettings.countX = this->parameters.arrangement.columnWidth;
				columnSettings.countY = this->parameters.arrangement.rows;
				columnSettings.flipped = this->parameters.arrangement.flipped;
			};

			for (int colIndex = 0; colIndex < this->parameters.arrangement.columns; colIndex++) {
				columnSettings.index = colIndex;
				auto column = make_shared<Column>(columnSettings);
				this->columns.push_back(column);
				column->init();
			}

			this->needsRebuildPanel = true;

			this->needsRebuildColumns = false;
		}

		//----------
		void
			Installation::populateInspector(ofxCvGui::InspectArguments& args)
		{
			auto inspector = args.inspector;

			inspector->addParameterGroup(this->parameters);
			inspector->addButton("Rebuild columns", [this]() {
				this->rebuildColumns();
				}, OF_KEY_RETURN)->setHeight(100.0f);

			this->massFWUpdate->addSubMenuToInsecptor(inspector, this->massFWUpdate);

			// Add simple pilot (draggable button
			{
				inspector->addTitle("Pilot all:", ofxCvGui::Widgets::Title::Level::H3);

				auto pilotButton = inspector->addButton("", []() {});
				pilotButton->setHeight(inspector->getWidth());

				// Add mouse action
				auto pilotButtonWeak = weak_ptr<ofxCvGui::Element>(pilotButton);
				pilotButton->onMouse += [this, pilotButtonWeak](ofxCvGui::MouseArguments& args) {
					if (this->columns.empty()) {
						return;
					}
					auto portals = this->columns.front()->getAllPortals();
					if (portals.empty()) {
						return;
					}

					auto pilotButton = pilotButtonWeak.lock();
					if (pilotButton->isMouseDown()) {
						auto firstPortal = portals.front();
						auto firstPilot = firstPortal->getPilot();

						auto position = args.localNormalized * 2.0f - 1.0f;

						// clamp to r<=1
						if (glm::length(position) > 1.0f) {
							position /= glm::length(position);
						}

						auto polar = firstPilot->positionToPolar(position);
						auto axes = firstPilot->polarToAxes(position);

						auto stepsA = firstPilot->axisToSteps(axes[0], 0);
						auto stepsB = firstPilot->axisToSteps(axes[1], 1);

						this->broadcast(MsgPack::object{
							{
								"m"
								, MsgPack::array {
									stepsA
									, stepsB
								}
							}
							}, true);
					}
					};
			}
		}

		//----------
		vector<shared_ptr<Column>>
			Installation::getAllColumns() const
		{
			return this->columns;
		}

		//----------
		shared_ptr<Column>
			Installation::getColumnByID(size_t columnID) const
		{
			if (columnID > this->columns.size()) {
				return shared_ptr<Column>();
			}
			return this->columns[columnID];
		}

		//----------
		vector<shared_ptr<Portal>>
			Installation::getAllPortals() const
		{
			vector<shared_ptr<Portal>> portals;
			for (const auto& column : this->columns) {
				auto columnPortals = column->getAllPortals();
				portals.insert(portals.end(), columnPortals.begin(), columnPortals.end());
			}

			return portals;
		}

		//----------
		shared_ptr<Portal>
			Installation::getPortalByTargetID(size_t columnID, Portal::Target target) const
		{
			auto column = this->getColumnByID(columnID);
			if (!column) {
				return shared_ptr<Portal>();
			}

			return column->getPortalByTargetID(target);
		}

		//----------
		glm::tvec2<size_t>
			Installation::getResolution() const
		{
			auto allColumns = this->getAllColumns();
			if (allColumns.empty()) {
				return { 0, 0 };
			}

			auto testColumn = allColumns.front();
			auto testPortals = testColumn->getAllPortals();

			return {
				this->columns.size() * testColumn->getCountX()
				, testColumn->getCountY()
			};
		}

		//----------
		void
			Installation::pollAll()
		{
			for (const auto& column : this->columns) {
				column->pollAll();
			}
		}

		//----------
		void
			Installation::broadcast(const msgpack11::MsgPack& message, bool collateable)
		{
			auto columns = this->getAllColumns();
			for (const auto& column : columns) {
				column->broadcast(message, collateable);
			}
		}

		//----------
		ofxCvGui::PanelPtr
			Installation::getPanel()
		{
			return this->panel;
		}

		//----------
		ofxCvGui::PanelPtr
			Installation::getMiniView()
		{
			auto panel = ofxCvGui::Panels::makeBlank();
			panel->onDraw += [this](ofxCvGui::DrawArguments& args) {
				if (this->columns.empty()) {
					return;
				}
				const auto& firstColumn = this->columns.front();

				if (firstColumn->getAllPortals().empty()) {
					return;
				}

				const auto countX = this->columns.size();
				const auto countY = firstColumn->getAllPortals().size();

				const auto cellSize = min(args.localBounds.width / countX, args.localBounds.height / countY);
				const auto radius = cellSize * 0.45 - 1;
				for (int i = 0; i < countX; i++) {
					const auto column = this->columns[i];
					const auto portals = column->getAllPortals();
					auto x = i * cellSize;

					for (int j = 0; j < countY; j++) {
						auto portal = portals[j];

						// We ignore flipped for now

						auto y = args.localBounds.height - (1 + j) * cellSize;
						auto targetPosition = portal->getPilot()->getPosition();
						auto livePosition = portal->getPilot()->getLivePosition();

						ofPushMatrix();
						{
							ofTranslate(x + cellSize / 2.0f, y + cellSize / 2.0f);
							ofPushStyle();
							{
								ofSetColor(80.0f);
								ofFill();
								ofDrawCircle(0, 0, radius);

								ofSetColor(255.0f);
								ofFill();
								ofDrawCircle(radius * targetPosition.x, radius * targetPosition.y, 2.0f);
								ofDrawLine({ 0, 0 }, radius * targetPosition);

								ofSetColor(100, 100, 200);
								ofDrawCircle(radius * livePosition.x, radius * livePosition.y, 1.0f);
								ofDrawLine({ 0, 0 }, radius * livePosition);

							}
							ofPopStyle();
						}
						ofPopMatrix();
					}
				}
				};

			return panel;
		}

		//----------
		void
			Installation::takeImage()
		{
			auto imageRendererModule = App::X()->getImageRenderer();
			const auto& pixels = imageRendererModule->getPixels();

			auto resolution = this->getResolution();
			if (resolution.x != pixels.getWidth() || resolution.y != pixels.getHeight()) {
				ofLogError("Installation::transmitImage") << "Resolution mismatch";
				return;
			}

			for (auto column : this->columns) {
				column->updatePositionsFromImage(pixels);
			}
		}

		//----------
		void
			Installation::transmitFrame()
		{
			// Choose between keyframe style (regular) or individual (on stale)

			switch (this->parameters.messaging.transmit.get()) {
			case ImageTransmit::Keyframe:
			{

				auto sendInterval = this->getTransmitKeyframeInterval();
				auto now = chrono::system_clock::now();
				auto doSend = this->lastTransmitKeyframe + sendInterval <= now;
				if (doSend) {
					for (auto column : this->columns) {
						column->transmitKeyframe();
					}
					this->lastTransmitKeyframe = chrono::system_clock::now();
				}
				break;
			}
			case ImageTransmit::Inidividual:
			{
				for (auto column : this->columns) {
					column->pushStale();
				}
				break;
			}
			case ImageTransmit::Disabled:
			{
				break;
			}
			}
		}

		//----------
		chrono::system_clock::duration
			Installation::getTransmitKeyframeInterval() const
		{
			return chrono::milliseconds((uint32_t)(this->parameters.messaging.periodS.get() * 1000.0f));
		}

		//----------
		int
			Installation::getTransmitKeyframeBatchSize() const
		{
			return this->parameters.messaging.keyframeBatchSize.get();
		}

		//----------
		bool
			Installation::getKeyframeVelocitiesEnabled() const
		{
			return this->parameters.messaging.keyframeVelocities.get();
		}

		//----------
		void
			Installation::rebuildPanel() 
		{
			this->panel->clear();

			// Actions (we don't support hot keys for the Installation broadcast actions)
			{
				this->panel->addTitle("Broadcast actions");

				auto buttonStack = this->panel->addHorizontalStack();
				// Special action for poll
				buttonStack->addButton("Poll", [this]() {
					this->pollAll();
					})->setDrawGlyph(u8"\uf059");

					// Add actions
					auto actions = Portal::getActions();
					for (const auto& action : actions) {
						// Max 6 elements per stack then start a new one
						if (buttonStack->getElements().size() >= 6) {
							buttonStack = this->panel->addHorizontalStack();
						}

						auto buttonAction = [this, action]() {
							this->broadcast(action.message, false);
							};

						auto button = buttonStack->addButton(action.caption, buttonAction);
						button->setDrawGlyph(action.icon);
					}
			}

			this->panel->addSpacer();

			// Add columns
			{
				this->panel->addTitle("Columns");

				auto stack = this->panel->addHorizontalStack();
				float colHeight = 10.0f;
				for (const auto& column : this->columns) {
					auto columnWeak = weak_ptr<Column>(column);
					auto button = make_shared<ofxCvGui::Widgets::SubMenuInspectable>("", column);

					// Custom draw
					{
						button->onDraw += [columnWeak](ofxCvGui::DrawArguments& args) {
							auto column = columnWeak.lock();
							if (ofxCvGui::isBeingInspected(column)) {
								ofPushStyle();
								{
									ofNoFill();
									ofDrawRectangle(args.localBounds);
								}
								ofPopStyle();
							}
							};
					}

					// Vertical stack of elements in the button
					float height = 0.0f;
					{

						auto verticalStack = make_shared<ofxCvGui::Widgets::VerticalStack>(ofxCvGui::Widgets::VerticalStack::Layout::UseElementHeight);
						{
							button->addChild(verticalStack);
							auto verticalStackWeak = weak_ptr<ofxCvGui::Element>(verticalStack);
							button->onBoundsChange += [verticalStackWeak](ofxCvGui::BoundsChangeArguments& args) {
								auto verticalStack = verticalStackWeak.lock();
								auto bounds = args.localBounds;
								bounds.width -= 60.0f;
								verticalStack->setBounds(bounds);
								};
						}

						// Title
						{
							auto element = make_shared<ofxCvGui::Widgets::Title>(column->getName());
							verticalStack->add(element);
							height += element->getHeight();
						}

						// RS485 connected
						{
							auto element = make_shared<ofxCvGui::Widgets::Indicator>("RS485", [columnWeak]() {
								auto column = columnWeak.lock();
								if (column->getRS485()->isConnected()) {
									return ofxCvGui::Widgets::Indicator::Status::Good;
								}
								else {
									return ofxCvGui::Widgets::Indicator::Status::Warning;
								}
								});
							verticalStack->add(element);
							height += element->getHeight();
						}

						// Outbox size
						{
							auto element = make_shared<ofxCvGui::Widgets::LiveValue<size_t>>("Outbox size", [columnWeak]() {
								auto column = columnWeak.lock();
								return column->getRS485()->getOutboxCount();
								});
							verticalStack->add(element);
							height += element->getHeight();
						}

						// Tx and Rx heartbeats
						{
							auto widgets = column->getRS485()->getWidgets();
							for (auto widget : widgets) {
								verticalStack->add(widget);
								height += widget->getHeight();
							}
						}

						// Preview of all elements
						{
							auto element = column->getMiniView(ofGetWidth() / this->columns.size() - 20.0f);
							verticalStack->add(element);
							height += element->getHeight();
						}

						// Spacer at bottom
						{
							auto element = make_shared<ofxCvGui::Widgets::Spacer>();
							verticalStack->add(element);
							height += element->getHeight();
						}
					}
					button->setHeight(height);
					colHeight = height;

					stack->add(button);
				}

				stack->setHeight(colHeight);
			}

			this->needsRebuildPanel = false;
		}
	}
}

File: RouterRS/crates/router-core/src/model/installation.rs
//! The Installation: owns all Columns, the arrangement, image->hardware
//! transmission, and broadcast actions.
//! Port of `Router/src/Modules/Hardware/Installation.*`.

use std::time::Instant;

use router_proto::commands::ActionKind;
use router_proto::Value;
use router_report::Reporter;

use crate::config::{ImageTransmit, InstallationConfig};
use crate::image::PixelsF32;

use super::column::{Column, ColumnSettings};

pub struct Installation {
    // arrangement parameters
    pub columns_count: usize,
    pub rows: usize,
    pub column_width: usize,
    /// Rows per panel, when a column is a stack of panels rather than a flat grid.
    pub panel_height: usize,
    pub flipped: bool,
    // messaging parameters
    pub transmit: ImageTransmit,
    pub period_s: f32,
    pub keyframe_batch_size: usize,
    pub keyframe_velocities: bool,
    // image parameter
    pub image_enabled: bool,

    pub columns: Vec<Column>,
    reporter: Reporter,
    last_transmit_keyframe: Instant,
}

impl Installation {
    pub fn new(reporter: Reporter) -> Self {
        let mut installation = Self {
            columns_count: 32,
            rows: 24,
            column_width: 1,
            panel_height: 0,
            flipped: false,
            transmit: ImageTransmit::Individual,
            period_s: 0.5,
            keyframe_batch_size: 8,
            keyframe_velocities: true,
            image_enabled: false,
            columns: Vec::new(),
            reporter,
            last_transmit_keyframe: Instant::now(),
        };
        installation.rebuild_columns();
        installation
    }

    pub fn from_config(config: &InstallationConfig, reporter: Reporter) -> Self {
        let mut installation = Self::new(reporter);
        installation.columns_count = config.arrangement.columns;
        installation.rows = config.arrangement.rows;
        installation.column_width = config.arrangement.column_width;
        installation.panel_height = config.arrangement.panel_height;
        installation.flipped = config.arrangement.flipped;
        installation.transmit = config.messaging.transmit;
        installation.period_s = config.messaging.period_s;
        installation.keyframe_batch_size = config.messaging.keyframe_batch_size;
        installation.keyframe_velocities = config.messaging.keyframe_velocities;
        installation.image_enabled = config.image_enabled;
        installation.rebuild_columns();

        // per-column settings (like C++: min(columns, config entries))
        let count = installation.columns.len().min(config.columns.len());
        for i in 0..count {
            installation.columns[i].apply_config(&config.columns[i]);
        }
        installation
    }

    pub fn rebuild_columns(&mut self) {
        self.columns = (0..self.columns_count)
            .map(|index| {
                Column::new(
                    ColumnSettings {
                        index,
                        count_x: self.column_width,
                        count_y: self.rows,
                        panel_height: self.panel_height,
                        flipped: self.flipped,
                    },
                    self.reporter.clone(),
                )
            })
            .collect();
    }

    /// Image resolution derived from the arrangement:
    /// (columns * countX, rows).
    pub fn resolution(&self) -> (usize, usize) {
        match self.columns.first() {
            Some(first) => (self.columns.len() * first.count_x, first.count_y),
            None => (0, 0),
        }
    }

    /// Per-tick: columns update, then frame transmission per the messaging
    /// mode. `pixels` is the rendered image when image sampling is enabled.
    pub fn update(&mut self, pixels: Option<&PixelsF32>) {
        if self.image_enabled {
            if let Some(pixels) = pixels {
                self.take_image(pixels);
            }
        }
        self.transmit_frame();

        let reporter = self.reporter.clone();
        for column in &mut self.columns {
            column.update(&reporter);
        }
    }

    /// `Installation::takeImage`: verify resolution and push pixel targets
    /// into every column.
    pub fn take_image(&mut self, pixels: &PixelsF32) {
        let (w, h) = self.resolution();
        if pixels.width != w || pixels.height != h {
            return;
        }
        for column in &mut self.columns {
            column.update_positions_from_image(pixels);
        }
    }

    /// `Installation::transmitFrame`.
    pub fn transmit_frame(&mut self) {
        match self.transmit {
            ImageTransmit::Keyframe => {
                if self.last_transmit_keyframe.elapsed().as_secs_f32() >= self.period_s {
                    self.last_transmit_keyframe = Instant::now();
                    let batch = self.keyframe_batch_size;
                    let velocities = self.keyframe_velocities;
                    for column in &mut self.columns {
                        column.transmit_keyframe(batch, velocities);
                    }
                }
            }
            ImageTransmit::Individual => {
                for column in &mut self.columns {
                    column.push_stale();
                }
            }
            ImageTransmit::Disabled => {}
        }
    }

    pub fn poll_all(&mut self) {
        for column in &mut self.columns {
            column.poll_all();
        }
    }

    pub fn broadcast(&self, body: &Value, collateable: bool) {
        for column in &self.columns {
            column.broadcast(body, collateable);
        }
    }

    pub fn broadcast_action(&mut self, action: ActionKind) {
        for column in &mut self.columns {
            column.broadcast_action(action);
        }
    }

    /// `homeHardwareAndZeroPositions`: broadcast the Home action 10 times,
    /// then reset every pilot.
    pub fn home_hardware_and_zero_positions(&mut self) {
        for _ in 0..10 {
            self.broadcast(&ActionKind::Home.body(), false);
        }
        for column in &mut self.columns {
            for portal in &mut column.portals {
                portal.pilot.reset();
            }
        }
    }

    pub fn column(&mut self, index: usize) -> Option<&mut Column> {
        self.columns.get_mut(index)
    }

    pub fn portal(&mut self, column_index: usize, target: u8) -> Option<&mut super::portal::Portal> {
        self.columns
            .get_mut(column_index)
            .and_then(|c| c.portal_by_target(target))
    }
}
